cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+            self._value = self._construct_value(None for _ in range)
+            self._range = self._construct_range(range)
+        else:
+            raise TypeError("must pass a value, range, or both")
+
+    @staticmethod
+    def _construct_value(value: Iterable[Any]) -> List[Any]:
+        """
+        Constructs the value portion of the array.
+
+        For overriding by subclasses.
+        Used by constructor when values are given
+        and by __setitem__ when setting a slice to an iterable.
+
+        Args:
+            value: Any iterable of values that are constructable into elements using :meth:`_construct_elements`
+
+        Returns:
+            A :class:`list` of elements.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the iterable's values prevent construction into a list of elements.
+        """
+        return list(value)
+
+    @staticmethod
+    def _construct_element(elem: Any) -> Any:
+        """
+        Constructs a single element of an array.
+
+        For overriding by subclasses.
+        Used by __setitem__ when setting a single element.
+
+        Args:
+            elem: Any object that the implementation can turn into an array element.
+
+        Returns:
+            An array element.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Constructs the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
+
+    @cached_property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self._range.left
+
+    @cached_property
+    def direction(self) -> str:
+        """``'to'`` if indexes are ascending, ``'downto'`` otherwise."""
+        return self._range.direction
+
+    @cached_property
+    def right(self) -> int:
+        """Rightmost index of the array."""
+        return self._range.right
+
+    @cached_property
+    def length(self):
+        """Length of the array."""
+        return self._range.length
+
+    @cached_property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
```

Comment:
Why not just call the attribute `self.range` in the first place? `cached_property` does exactly that anyway, but defers it to first access.

Reply:
Read-only access. But maybe that's not super important.

Reply:
This doesn't make it readonly anyway - try it and see.

Reply:
`@property` instead of `@cached_property` would.

Reply:
Yeah. At one point it was doing some non-trivial stuff, but it's not necessary anymore. I just removed all uses of `cached_property` and replaced it with `property`. It also means the code in `_py_compat` isn't exactly correct.